rfcs/2021-03-26-6517-end-to-end-acknowledgement.md

```diff
@@ -0,0 +1,344 @@
+separate channel for each sender or batch, as convenient for the source
+protocol, and then `await` on the receiver to collect the finalization
+events. Since the size of each batch is known in advance, this may be a
+bounded channel with the length fixed to the size of the source batch.
+When the last event in the batch is dropped, the sender will be closed
+and the receiver will be signalled that all events are completed.
+
+Since this is making the parameter list for `trait SourceConfig::build`
+increasingly unweidly, a new context structure will be introduced to
+carry the data.
+
+```rust
+struct SourceContext<'a> {
+    name: &'a str,
+    identifier: Box<str>,
+    globals: &'a GlobalOptions,
+    shutdown: ShutdownSignal,
+    out: Pipeline,
+}
+
+trait SourceConfig {
+    async fn build(&self, context: SourceContext<'_>) -> crate::Result<sources::Source>;
+}
+```
+
+### Sink configuration
+
+A new global configuration setting will be added to all sources,
+flagging one or more sinks as "authoritative". This setting will change
+the behavior of finalization as described above. When an event is
+delivered to an authoritative sink, its finalization status is
+immediately delivered to all sources of the event. In order to prevent
+another status from being delivered at a later point, the status of the
+event is changed to "no-op", the same as if it was merged into another
+event.
+
+```rust
+struct SinkOuter {
+    // … existing fields …
+
+    #[serde(default)]
+    authoritative: bool,
+}
+```
+
+## Doc-level Proposal
+
+The following additional source configuration will be added:
+
+```toml
+# Global enable option
+acknowledgements = true
+
+[sources.my_source_id]
+  # Per-source enable option
+  acknowledgements = true
+```
+
+We would also need to document when acknowledgement happens for each source.
+
+## Rationale
+
+The above structure provides for several considerations:
+
+1.  This the minimum amount of data that can be added to the metadata to
+    fully support this feature, amounting to a single shared reference
+    as the `Option` is optimized into the `Arc`.
+
+2.  The use of `Arc` reference counting for finalization prevents events
+    from "escaping" without providing a status indication.
+
+3.  If a source does not need or is not configured to require
+    finalization, it will not contribute to the list of sources and so
+    has no additional event overhead, and no additional allocations when
+    the event is created.
+
+4.  Sending the finalization status to the source does not require any
+    lookups or topology traversal.
+
+5.  No additional work is required to handle dropped sources due to
+    topology reconfiguration, other than the expected checking for a
+    closed channel when sending.
+
+## Drawbacks
+
+1.  This adds a base size overhead to each event, even for
+    configurations that do not support or require end-to-end
+    acknowledgement.
```

Comment:
I wonder if we could make this pay for what you use by making our event metadata generic. So, the metadata becomes 

```
struct EventMetadata<F: Finalizer> {
    // … existing fields …
    finalizer: Option<Arc<F>>,
}
```

and for any structure that doesn't need to ack we make the `Finalizer` zero-sized. This would preclude topologies that mix ack vs. not ack. I'm also late to this RFC so that may have already been discussed.

Reply:
I'm not sure how this would work in practice that would actually change the size requirements. The metadata is stored as a field in either `struct LogEvent` or `struct Metric`, which are not generic. So, it would have to be in an enum, eliminating the size benefit, or make `enum Event` generic, and that way lies madness.

Reply:
> or make enum Event generic, and that way lies madness.

Yes, you're quite right.